### PR TITLE
txn: Add lock-with-conflict support to acquire_pessimistic_lock

### PR DESCRIPTION
**Pull request by MyonKeminta**
Signed-off-by: MyonKeminta <MyonKeminta@users.noreply.github.com>

### What is changed and how it works?

Issue Number: Ref #13298 

What's Changed:

```commit-message
Add `lock_with_conflict` support to  `acquire_pessimistic_lock`, and it's currently always disabled.

Side changes:
Updated the type for holding the results of `acquire_pessimistic_lock` (removed `PessimisticLockRes` and added new `PessimisticLockResults` and `PessimisticLockKeyResult`), and other necessary changes to adapt the new type.
```

### Check List 

Tests 

- Unit test

Side effects

- Performance regression
    - Consumes more CPU

### Release note 

```release-note
None
```

### CODE REVIEW

File: src/storage/mod.rs [hunk 1]
```diff
@@ -3253,10 +3260,56 @@ pub mod test_util {
...
+            lhs: &PessimisticLockKeyResult,
+            rhs: &PessimisticLockKeyResult,
+        ) -> bool {
+            match (lhs, rhs) {
+                (PessimisticLockKeyResult::Empty, PessimisticLockKeyResult::Empty) => true,
+                (PessimisticLockKeyResult::Value(l), PessimisticLockKeyResult::Value(r))
+                    if l == r =>
+                {
+                    true
+                }
```
Comment (sticnarf) [inline src/storage/mod.rs:3275]: Maybe it is cleaner to write it like 
```suggestion
                     => l == r
```

File: src/storage/mod.rs [hunk 2]
```diff
@@ -3253,10 +3260,56 @@ pub mod test_util {
 
     pub fn expect_pessimistic_lock_res_callback(
         done: Sender<i32>,
-        pessimistic_lock_res: PessimisticLockRes,
-    ) -> Callback<Result<PessimisticLockRes>> {
-        Box::new(move |res: Result<Result<PessimisticLockRes>>| {
-            assert_eq!(res.unwrap().unwrap(), pessimistic_lock_res);
+        pessimistic_lock_res: PessimisticLockResults,
+    ) -> Callback<Result<PessimisticLockResults>> {
+        fn key_res_matches_ignoring_error_content(
```
Comment (sticnarf) [inline src/storage/mod.rs:3265]: Is there any special reason why we don't compare the error content?

**Review by cfzjywxk — APPROVED** [review]
LGTM

**Comment by MyonKeminta** [comment]
I gave up about the feature branch since these split PRs are not expected to affect the original behaviors, and feature branch doesn't look effective to me. I'll re-target these PRs to master.

File: src/storage/txn/actions/acquire_pessimistic_lock.rs [hunk 3]
```diff
@@ -10,18 +10,26 @@ use crate::storage::{
...
-/// of the key (possibly `None`). When `need_value` is not set but
-/// `need_check_existence` is set, the first return value will be an empty value
-/// (`Some(vec![])`) if the key exists before or `None` if not. If neither
-/// `need_value` nor `need_check_existence` is set, the first return value is
-/// always `None`.
+/// When `need_value` is set, the first return value will be
+/// `PessimisticLockKeyResult::Value`. When `need_value` is not set but
+/// `need_check_existence` is set, the first return value will be
+/// `PessimisticLockKeyResult::Existence`. If neither `need_value` nor
+/// `need_check_existence` is  set, the first return value will be
```
Comment (sticnarf) [inline src/storage/txn/actions/acquire_pessimistic_lock.rs:24]:
```suggestion
/// `need_check_existence` is set, the first return value will be
```

File: src/storage/txn/actions/acquire_pessimistic_lock.rs [hunk 4]
```diff
@@ -116,6 +113,16 @@ pub fn acquire_pessimistic_lock<S: Snapshot>(
             }
             .into());
         }
+
+        let locked_with_conflict_ts =
```
Comment (sticnarf) [inline src/storage/txn/actions/acquire_pessimistic_lock.rs:117]: Better to comment that this is for idempotence.

**Review by sticnarf — COMMENTED** [review]
Mostly lgtm

File: src/storage/txn/actions/acquire_pessimistic_lock.rs [hunk 5]
```diff
@@ -116,6 +113,32 @@ pub fn acquire_pessimistic_lock<S: Snapshot>(
...
+                // for_update_ts, and the current request has
+                // `allow_lock_with_conflict` set, we must consider
+                // these possibilities:
+                // * If a previous request successfully locked the key with conflict, but the
+                //   response is lost due to some errors such as RPC failures. In this case, we
+                //   return like the current request's result is locked_with_conflict, for
+                //   idempotency concern.
+                // * The key is locked by a newer request with larger for_update_ts, and the
+                //   current request is stale. We can't distinguish this case with the above
+                //   one, but we don't need to handle this case since no one wood need the
```
Comment (sticnarf) [inline src/storage/txn/actions/acquire_pessimistic_lock.rs:129]: wood -> would?

Comment (MyonKeminta) [reply src/storage/txn/actions/acquire_pessimistic_lock.rs:129]: ouch

**Review by sticnarf — APPROVED** [review]
LGTM

**Pull request merged**